File: clients/typescript/scripts/check-readme-hero.mjs
// Enforces that the repo README's "hero" TypeScript snippet is byte-identical to the type-checked example in
// examples/hero.ts — so the snippet we show the world always compiles against the real client. The example is
// type-checked by `pnpm typecheck`; this script proves the README matches it. Run via `pnpm example`.
import { readFile } from "node:fs/promises";

const README = new URL("../../../README.md", import.meta.url); // repo root README
const HERO = new URL("../examples/hero.ts", import.meta.url);

const START = "// >>> readme-hero";
const END = "// <<< readme-hero";

function extractRegion(src) {
  const s = src.indexOf(START);
  const e = src.indexOf(END, s + START.length);
  if (s === -1 || e === -1) {
    throw new Error("examples/hero.ts is missing the readme-hero markers");
  }
  return src.slice(s + START.length, e).trim();
}

function extractTsBlocks(md) {
  const blocks = [];
  const re = /```ts\r?\n([\s\S]*?)```/g;
  let m;
  while ((m = re.exec(md)) !== null) blocks.push(m[1].trim());
  return blocks;
}

const [md, hero] = await Promise.all([readFile(README, "utf8"), readFile(HERO, "utf8")]);
const want = extractRegion(hero);
const blocks = extractTsBlocks(md);

if (blocks.some((b) => b === want)) {
  console.log("ok: the README hero snippet matches examples/hero.ts");
  process.exit(0);
}

console.error(
  "The README hero snippet does not match examples/hero.ts.\n" +
    "Update the fenced ```ts block in README.md to exactly:\n\n" +
    want +
    "\n",
);
process.exit(1);

File: clients/python/scripts/check_readme_hero.py
"""Assert the package README's `python` hero block is byte-identical to the marked region of examples/hero.py,
so the shipped snippet always type-checks against the real client. Run via `python scripts/check_readme_hero.py`."""

from __future__ import annotations

import re
import sys
from pathlib import Path

_ROOT = Path(__file__).parent.parent
_README = _ROOT / "README.md"
_HERO = _ROOT / "examples" / "hero.py"
_START = "# >>> readme-hero"
_END = "# <<< readme-hero"


def _region(text: str) -> str:
    start = text.index(_START)
    end = text.index(_END, start + len(_START))
    return text[start + len(_START) : end].strip()


def _python_blocks(md: str) -> list[str]:
    return [m.strip() for m in re.findall(r"```python\r?\n(.*?)```", md, re.DOTALL)]


def main() -> int:
    want = _region(_HERO.read_text(encoding="utf-8"))
    if want in _python_blocks(_README.read_text(encoding="utf-8")):
        print("ok: the README hero snippet matches examples/hero.py")
        return 0
    print(
        "The README hero snippet does not match examples/hero.py. Update the ```python block to:\n",
        file=sys.stderr,
    )
    print(want, file=sys.stderr)
    return 1


if __name__ == "__main__":
    sys.exit(main())
